Dockerfile

```diff
@@ -1,85 +1,52 @@
-RUN apt-get update && apt-get install -y git
+# xx is a helper for cross-compilation
+FROM --platform=$BUILDPLATFORM tonistiigi/xx@sha256:56b19a5fb89b99195ec494d59ad34370d14540858c1f56c560ec1e7f2d1c177f AS xx
+
+FROM --platform=$BUILDPLATFORM ${ALPINE_BASE} AS src
+RUN apk add --no-cache git patch
 ARG QEMU_VERSION
 ARG QEMU_REPO
 WORKDIR /src
 RUN git clone $QEMU_REPO && \
+  git clone --depth 1 -b 3.14-stable https://github.com/alpinelinux/aports.git && \
   cd qemu && \
   git checkout $QEMU_VERSION && \
+  for f in  ../aports/community/qemu/*.patch; do patch -p1 < $f; done && \
   scripts/git-submodule.sh update \
   ui/keycodemapdb \
   tests/fp/berkeley-testfloat-3 \
   tests/fp/berkeley-softfloat-3 \
   dtc slirp
 
-FROM --platform=$BUILDPLATFORM debian:buster AS qemu
-
-RUN apt-get update && \
-  apt-get install -y \
-  dpkg-dev \
-  git \
-  ninja-build \
-  pkg-config \
-  python3 \
-  python3-setuptools
-
-WORKDIR /qemu
-
-ARG TARGETPLATFORM
-
-ENV PATH=/qemu/install-scripts:$PATH
-RUN --mount=target=./install-scripts,src=scripts \
-  TARGETPLATFORM=${TARGETPLATFORM} cross.sh install gcc libglib2.0-dev | sh
-
-FROM qemu AS base-amd64
-FROM qemu AS base-arm64
-FROM qemu AS base-ppc64le
-FROM qemu AS base-s390x
-FROM qemu AS base-armv7
-FROM qemu AS base-armv6
-FROM qemu AS base-386
-FROM qemu AS base-mips64le
-
-FROM tonistiigi/debian:riscv AS riscv-libglibc
-RUN apt-get update && apt-get install -y libglib2.0-dev
-
-RUN for f in $(dpkg-query -L zlib1g-dev libglib2.0-dev libpcre3-dev libglib2.0-0 libpcre3); do [ ! -d $f ] && echo $f; done > /tmp/list
-RUN mkdir -p /out && tar cvf /out/libglibc.tar -T /tmp/list
-
-FROM tonistiigi/xx:riscv-toolchain AS base-riscv64
-RUN apt-get update && \
-  apt-get install -y \
-  dpkg-dev \
-  git \
-  ninja-build \
-  pkg-config \
-  python3 \
-  python3-setuptools
+FROM --platform=$BUILDPLATFORM ${ALPINE_BASE} AS base
+RUN apk add --no-cache git clang lld python3 llvm make ninja pkgconfig glib-dev gcc musl-dev perl bash
+COPY --from=xx / /
 ENV PATH=/qemu/install-scripts:$PATH
 WORKDIR /qemu
 
-RUN --mount=from=riscv-libglibc,target=/riscv-libglibc,src=out \
-  mkdir -p /tmp/out && tar xvf /riscv-libglibc/libglibc.tar -C /tmp/out && \
-  cp -a /tmp/out/usr/include/* /usr/riscv64-linux-gnu/include/ && \
-  cp -a /tmp/out/usr/lib/riscv64-linux-gnu/* /usr/riscv64-linux-gnu/lib/ && \
-  cp -a /tmp/out/usr/lib/* /usr/riscv64-linux-gnu/lib/ && \
-  ln -s /usr/riscv64-linux-gnu /usr/riscv64-buildroot-linux-gnu
-ENV CROSS_PREFIX=riscv64-buildroot-linux-gnu
+ARG TARGETPLATFORM
+RUN xx-apk add musl-dev gcc glib-dev glib-static linux-headers zlib-static
+RUN set -e; \
+  [ "$(xx-info arch)" = "ppc64le" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple; \
+  [ "$(xx-info arch)" = "386" ] && XX_CC_PREFER_LINKER=ld xx-clang --setup-target-triple; \
+  true
 
-FROM base-$TARGETARCH$TARGETVARIANT AS base
 
 FROM base AS build
 ARG TARGETPLATFORM
-ARG QEMU_VERSION
+ARG QEMU_VERSION QEMU_TARGETS
```

Comment:
Should not be?:

```suggestion
ARG QEMU_VERSION
ARG QEMU_TARGETS
```

Reply:
What's the difference?

Reply:
I didn't think this syntax could work but apparently it does!